pkg/logql/rangemapper.go

```diff
@@ -0,0 +1,278 @@
+		// sum(x) -> sum(sum(x offset splitByInterval1) ++ sum(x offset splitByInterval2)...)
+		return &VectorAggregationExpr{
+			Left:      m.mapSampleExpr(expr, rangeInterval),
+			Grouping:  expr.Grouping,
+			Params:    expr.Params,
+			Operation: expr.Operation,
+		}, nil
+
+	case OpTypeCount:
+		// count(x) -> sum(count(x offset splitByInterval1) ++ count(x offset splitByInterval2)...)
+		return &VectorAggregationExpr{
+			Left:      m.mapSampleExpr(expr, rangeInterval),
+			Grouping:  expr.Grouping,
+			Operation: OpTypeSum,
+		}, nil
+	case OpTypeMax:
+		// max(x) -> max(max(x offset splitByInterval1) ++ max(x offset splitByInterval2)...)
+		return &VectorAggregationExpr{
+			Left:      m.mapSampleExpr(expr, rangeInterval),
+			Grouping:  expr.Grouping,
+			Operation: expr.Operation,
+		}, nil
+	case OpTypeMin:
+		// min(x) -> min(min(x offset splitByInterval1) ++ min(x offset splitByInterval2)...)
+		return &VectorAggregationExpr{
+			Left:      m.mapSampleExpr(expr, rangeInterval),
+			Grouping:  expr.Grouping,
+			Operation: expr.Operation,
+		}, nil
+	default:
+		// this should not be reachable. If an operation is shardable it should
+		// have an optimization listed.
+		level.Warn(util_log.Logger).Log(
+			"msg", "unexpected operation which appears shardable, ignoring",
+			"operation", expr.Operation,
+		)
+		return expr, nil
+	}
+}
+
+func (m RangeVectorMapper) mapRangeAggregationExpr(expr *RangeAggregationExpr) SampleExpr {
+	// TODO: In case expr is non-splittable, can we attempt to shard a child node?
+
+	rangeInterval := getRangeInterval(expr)
+
+	// in case the interval is smaller than the configured split interval,
+	// don't split it.
+	// TODO: what if there is another internal expr with an interval that can be split?
+	if rangeInterval <= m.splitByInterval {
+		return expr
+	}
+
+	if isSplittableByRange(expr) {
+		switch expr.Operation {
+		case OpRangeTypeBytes, OpRangeTypeBytesRate, OpRangeTypeCount, OpRangeTypeRate, OpRangeTypeSum:
+			return &VectorAggregationExpr{
+				Left:      m.mapSampleExpr(expr, rangeInterval),
+				Grouping:  &Grouping{},
+				Operation: OpTypeSum,
+			}
+		case OpRangeTypeMin:
+			// min_over_time support grouping
+			grouping := expr.Grouping
+			if grouping == nil {
+				grouping = &Grouping{}
+			}
+			return &VectorAggregationExpr{
+				Left:      m.mapSampleExpr(expr, rangeInterval),
+				Grouping:  grouping,
+				Operation: OpTypeMin,
+			}
+		case OpRangeTypeMax:
+			// max_over_time support grouping
+			grouping := expr.Grouping
+			if grouping == nil {
+				grouping = &Grouping{}
+			}
+			return &VectorAggregationExpr{
+				Left:      m.mapSampleExpr(expr, rangeInterval),
+				Grouping:  grouping,
+				Operation: OpTypeMax,
+			}
+		}
+	}
+
+	return expr
+}
+
+func isSplittableByRange(expr SampleExpr) bool {
+	switch e := expr.(type) {
+	case *VectorAggregationExpr:
+		_, ok := SplittableVectorOp[e.Operation]
+		return ok && isSplittableByRange(e.Left)
+	case *BinOpExpr:
+		return true
```

Comment:
Why does this return `true` when the `Vector` and `Range` options may not?

Reply:
This function will be called in a recursive call, so we want to focus on splitting vector and range vector aggregation expressions. Therefore, a binary operation is considered splittable, because we want to check if their operators are equally splittable.